File: extracted_app/GolfStatTracker/client/src/lib/golfCourseApi.ts
import { GolfCourse, SearchResult, TeeBox, Hole } from '../types/golfCourseApi';

// Using our server proxy instead of calling the Golf Course API directly
const API_PROXY_URL = '/api/golf-api';

/**
 * Search for golf courses by name
 * @param query The search term
 * @returns A promise with search results
 */
export async function searchGolfCourses(query: string): Promise<SearchResult> {
  try {
    const response = await fetch(`${API_PROXY_URL}/search?search_query=${encodeURIComponent(query)}`, {
      method: 'GET',
    });

    if (!response.ok) {
      let errorMessage = `Golf course search failed: ${response.statusText}`;
      try {
        const errorData = await response.json();
        console.error('API error:', errorData);
        if (errorData.message) {
          errorMessage = errorData.message;
        }
      } catch (e) {
        // If response is not JSON, just use the status text
      }
      throw new Error(errorMessage);
    }

    const data = await response.json();
    
    // Log the search results for debugging
    if (data.courses && Array.isArray(data.courses)) {
      console.log(`API returned ${data.courses.length} courses before filtering`);
      
      // Don't filter out US courses for now (enabling all results)
      // Just log what countries are available
      const countries = new Set();
      data.courses.forEach((course: GolfCourse) => {
        if (course.location && course.location.country) {
          countries.add(course.location.country);
        }
      });
      console.log('Available countries:', Array.from(countries));
    }
    
    return data;
  } catch (error) {
    console.error('Error searching for golf courses:', error);
    return { courses: [] };
  }
}

/**
 * Get detailed information about a specific golf course
 * @param id The golf course ID
 * @returns A promise with the course details
 */
export async function getGolfCourseById(id: number): Promise<GolfCourse> {
  console.log(`Fetching detailed info for course ID: ${id}`);
  
  try {
    const response = await fetch(`${API_PROXY_URL}/courses/${id}`, {
      method: 'GET',
      headers: {
        'Cache-Control': 'no-cache',
        'Pragma': 'no-cache'
      }
    });

    if (!response.ok) {
      let errorMessage = `Failed to get golf course details: ${response.statusText}`;
      try {
        const errorData = await response.json();
        console.error('API error:', errorData);
        if (errorData.message) {
          errorMessage = errorData.message;
        }
      } catch (e) {
        // If response is not JSON, just use the status text
      }
      throw new Error(errorMessage);
    }

    const courseData = await response.json();
    
    // Log the course data for debugging
    console.log(`Retrieved data for ${courseData.club_name || 'Unknown Course'}`);
    console.log(`Course has tee boxes:`, !!courseData.tees);
    
    // Normalize the course data to ensure it has all required fields
    const normalizedCourse: GolfCourse = {
      id: courseData.id,
      club_name: courseData.club_name || 'Unknown Club',
      course_name: courseData.course_name || 'Main Course',
      location: courseData.location || {},
      holes: courseData.holes || [],
      tees: courseData.tees || { male: [], female: [] }
    };
    
    return normalizedCourse;
  } catch (error) {
    console.error('Error getting golf course details:', error);
    // Return a default course with the requested ID as a fallback
    throw new Error(`Could not load course details: ${error}`);
  }
}

/**
 * Get a list of tee boxes available for a course
 * @param course The golf course object
 * @returns A formatted list of tee boxes with gender indication
 */
export function getFormattedTeeBoxes(course: GolfCourse): { label: string; value: string; data: TeeBox }[] {
  const teeBoxes: { label: string; value: string; data: TeeBox }[] = [];
  
  console.log('Formatting tee boxes for course:', course.club_name);

  // Log the raw tees object to debug
  console.log('Raw tees data:', JSON.stringify(course.tees || {}));

  try {
    // Add male tee boxes
    if (course.tees?.male && Array.isArray(course.tees.male)) {
      console.log(`Found ${course.tees.male.length} men's tee boxes`);
      course.tees.male.forEach((tee: TeeBox, index) => {
        // Validate tee box data
        if (!tee.tee_name) {
          tee.tee_name = `Men's Tee ${index + 1}`;
        }
        
        teeBoxes.push({
          label: `${tee.tee_name} (Men)`,
          value: `male-${tee.tee_name}`,
          data: tee
        });
      });
    } else {
      console.log('No men\'s tee boxes found or data is malformed');
    }

    // Add female tee boxes
    if (course.tees?.female && Array.isArray(course.tees.female)) {
      console.log(`Found ${course.tees.female.length} women's tee boxes`);
      course.tees.female.forEach((tee: TeeBox, index) => {
        // Validate tee box data
        if (!tee.tee_name) {
          tee.tee_name = `Women's Tee ${index + 1}`;
        }
        
        teeBoxes.push({
          label: `${tee.tee_name} (Women)`,
          value: `female-${tee.tee_name}`,
          data: tee
        });
      });
    } else {
      console.log('No women\'s tee boxes found or data is malformed');
    }
    
    // If no tee boxes were found, create a default one
    if (teeBoxes.length === 0 && course.holes && Array.isArray(course.holes) && course.holes.length > 0) {
      console.log('No tee boxes found, creating a default tee box from hole data');
      
      // Create a default tee box using the course holes
      const defaultTeeBox: TeeBox = {
        tee_name: 'Default',
        tee_color: 'White',
        tee_gender: 'male',
        par_total: course.holes.reduce((sum: number, hole: Hole) => sum + (hole.par || 4), 0),
        total_yards: course.holes.reduce((sum: number, hole: Hole) => sum + (hole.yardage || 0), 0),
        course_rating: 72,
        slope_rating: 113,
        number_of_holes: course.holes.length,
        holes: course.holes
      };
      
      teeBoxes.push({
        label: 'Default Tees (Men)',
        value: 'male-Default',
        data: defaultTeeBox
      });
    }
  } catch (error) {
    console.error('Error formatting tee boxes:', error);
  }

  console.log(`Returning ${teeBoxes.length} formatted tee boxes`);
  return teeBoxes;
}

/**
 * Convert API course and tee box data to local course data format
 */
export function convertToLocalCourse(course: GolfCourse, selectedTeeBox: TeeBox) {
  return {
    id: course.id,
    name: `${course.club_name} - ${course.course_name}`,
    city: course.location?.city || null,
    state: course.location?.state || null,
    country: course.location?.country || null,
    numHoles: selectedTeeBox.number_of_holes || 18,
    holeDetails: selectedTeeBox.holes || []
  };
}

/**
 * Format hole details for use in the application
 */
export function formatHoleDetails(holes: Hole[], courseId: number, roundId: number): any[] {
  return holes.map((hole, index) => ({
    roundId,
    holeNumber: index + 1,
    par: hole.par,
    distance: hole.yardage,
    score: null,
    fairwayHit: null,
    greenInRegulation: null,
    numPutts: null,
    numPenalties: null,
    upAndDown: null,
    sandSave: null,
    strokesGained: null
  }));
}

File: lib/models/golf_course_api.dart
// Golf Course API models
// Based on the original web application's types/golfCourseApi.ts

class GolfCourse {
  final int id;
  final String clubName;
  final String courseName;
  final Location location;
  final List<Hole> holes;
  final TeeBoxes tees;

  GolfCourse({
    required this.id,
    required this.clubName,
    required this.courseName,
    required this.location,
    required this.holes,
    required this.tees,
  });

  factory GolfCourse.fromJson(Map<String, dynamic> json) {
    // Safely convert id to int, handling null or non-int values
    int id = 0;
    try {
      if (json['id'] != null) {
        if (json['id'] is int) {
          id = json['id'];
        } else if (json['id'] is String) {
          id = int.tryParse(json['id']) ?? 0;
        }
      }
    } catch (e) {
      debugPrint('Error parsing course id: $e');
    }
    
    return GolfCourse(
      id: id,
      clubName: json['club_name'] ?? 'Unknown Club',
      courseName: json['course_name'] ?? 'Main Course',
      location: Location.fromJson(json['location'] ?? {}),
      holes: _parseHoles(json),
      tees: TeeBoxes.fromJson(json['tees'] ?? {'male': [], 'female': []}),
    );
  }
  
  // Helper method to safely parse holes
  static List<Hole> _parseHoles(Map<String, dynamic> json) {
    try {
      if (json['holes'] is List) {
        return (json['holes'] as List)
            .map((h) => Hole.fromJson(h is Map<String, dynamic> ? h : {}))
            .toList();
      }
    } catch (e) {
      debugPrint('Error parsing holes: $e');
    }
    return [];
  }

  factory GolfCourse.empty() {
    return GolfCourse(
      id: 0,
      clubName: '',
      courseName: '',
      location: Location(city: '', state: '', country: ''),
      holes: [],
      tees: TeeBoxes(male: [], female: []),
    );
  }
}

class SearchResult {
  final List<GolfCourse> courses;
  final bool missingApiKey;

  SearchResult({
    required this.courses,
    this.missingApiKey = false,
  });

  factory SearchResult.fromJson(Map<String, dynamic> json) {
    return SearchResult(
      courses: (json['courses'] as List<dynamic>?)
              ?.map((c) => GolfCourse.fromJson(c))
              .toList() ?? [],
      missingApiKey: json['missingApiKey'] == true,
    );
  }

  factory SearchResult.empty() {
    return SearchResult(courses: []);
  }
}

class Location {
  final String? city;
  final String? state;
  final String? country;

  Location({
    this.city,
    this.state,
    this.country,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      city: json['city'],
      state: json['state'],
      country: json['country'],
    );
  }
}

class Hole {
  final int holeNumber;
  final int par;
  final int yardage;
  final int? handicap;

  Hole({
    required this.holeNumber,
    required this.par,
    required this.yardage,
    this.handicap,
  });

  factory Hole.fromJson(Map<String, dynamic> json) {
    // Safely parse integer values with fallbacks
    int holeNumber = 0;
    int par = 4;
    int yardage = 0;
    int? handicap;
    
    try {
      // Hole number
      if (json['hole_num'] != null) {
        if (json['hole_num'] is int) {
          holeNumber = json['hole_num'];
        } else if (json['hole_num'] is String) {
          holeNumber = int.tryParse(json['hole_num']) ?? 0;
        }
      }
      
      // Par
      if (json['par'] != null) {
        if (json['par'] is int) {
          par = json['par'];
        } else if (json['par'] is String) {
          par = int.tryParse(json['par']) ?? 4;
        }
      }
      
      // Yardage
      if (json['yardage'] != null) {
        if (json['yardage'] is int) {
          yardage = json['yardage'];
        } else if (json['yardage'] is String) {
          yardage = int.tryParse(json['yardage']) ?? 0;
        }
      }
      
      // Handicap
      if (json['handicap'] != null) {
        if (json['handicap'] is int) {
          handicap = json['handicap'];
        } else if (json['handicap'] is String) {
          handicap = int.tryParse(json['handicap']);
        }
      }
    } catch (e) {
      debugPrint('Error parsing hole data: $e');
    }
    
    return Hole(
      holeNumber: holeNumber,
      par: par,
      yardage: yardage,
      handicap: handicap,
    );
  }
}

class TeeBoxes {
  final List<TeeBox> male;
  final List<TeeBox> female;

  TeeBoxes({
    required this.male,
    required this.female,
  });

  factory TeeBoxes.fromJson(Map<String, dynamic> json) {
    List<TeeBox> maleTees = [];
    List<TeeBox> femaleTees = [];
    
    try {
      if (json['male'] is List) {
        maleTees = (json['male'] as List)
            .map((t) => TeeBox.fromJson(t is Map<String, dynamic> ? t : {}))
            .toList();
      }
      
      if (json['female'] is List) {
        femaleTees = (json['female'] as List)
            .map((t) => TeeBox.fromJson(t is Map<String, dynamic> ? t : {}))
            .toList();
      }
    } catch (e) {
      debugPrint('Error parsing tee boxes: $e');
    }
    
    return TeeBoxes(
      male: maleTees,
      female: femaleTees,
    );
  }
}

class TeeBox {
  final String teeName;
  final String? teeColor;
  final String teeGender;
  final int parTotal;
  final int totalYards;
  final double? courseRating;
  final int? slopeRating;
  final int numberOfHoles;
  final List<Hole>? holes;

  TeeBox({
    required this.teeName,
    this.teeColor,
    required this.teeGender,
    required this.parTotal,
    required this.totalYards,
    this.courseRating,
    this.slopeRating,
    required this.numberOfHoles,
    this.holes,
  });

  factory TeeBox.fromJson(Map<String, dynamic> json) {
    return TeeBox(
      teeName: json['tee_name'] ?? 'Default',
      teeColor: json['tee_color'],
      teeGender: json['tee_gender'] ?? 'male',
      parTotal: json['par_total'] ?? 72,
      totalYards: json['total_yards'] ?? 6000,
      courseRating: json['course_rating'] != null ? 
                   double.tryParse(json['course_rating'].toString()) : 72.0,
      slopeRating: json['slope_rating'],
      numberOfHoles: json['number_of_holes'] ?? 18,
      holes: (json['holes'] as List<dynamic>?)
              ?.map((h) => Hole.fromJson(h))
              .toList(),
    );
  }
}

// Class for formatting tee boxes with selection data
class FormattedTeeBox {
  final String label;
  final String value;
  final TeeBox data;

  FormattedTeeBox({
    required this.label,
    required this.value,
    required this.data,
  });
}